### DISCUSSION TITLE
How to jump to next/previous error/warning/hint (`vim.lsp`)?

### DISCUSSION THREAD

**Discussion by Andrew15-5** [neovim/neovim#25588] [Q&A] [answered]
I switched from coc.nvim where I used to do `]g` & `[g` to jump to errors, but now that I switched to nvim-cmp + nvim-lspconfig, I can't find this most basic and one of the most important parts of developing in IDE. I found https://github.com/neovim/neovim/issues/12052#issuecomment-726492370, which says that I can do `vim.lsp.diagnostic.goto_next()`. I updated to https://github.com/neovim/neovim/commit/d772f697a281ce9c58bf933997b87c7f27428a60 which should work (since it is 0.9.4), but it doesn't! Because the function is `nil`, i.e., doesn't exist. There are also answers with `:lnext` and `:cnext`, which rely on "lists", but all I get is `E776: No location list`. But LSP server produces visible errors on specific lines, and yet I can't jump to them. Please someone help!

**Comment by clason** [comment]
Please read the documentation: `:h vim.diagnostic`. Your information is years out of date.

**Reply by Andrew15-5** [reply, marked as answer]
I started reading from the beginning and it was so complicated (I just want to jump to an error!). I was following the steps and kind of randomly entered the `goto_next()` function and was surprised as to why it suddenly started to exist (and it did jump to the next error). When I looked closely in excitement I saw `lua vim.diagnostic.goto_next()`. So, both jump commands are under `vim.diagnostic` instead of `vim.lsp.diagnostic`.

```lua
vim.keymap.set("n", "]g", vim.diagnostic.goto_next)
vim.keymap.set("n", "[g", vim.diagnostic.goto_prev)
```

**Comment by MarcWeber** [comment]
IDEA works 1) go to errors 2) then jump to hints. I didn't find this useful behavior looking at neovim's help and goto_next(). This seems to work for me:
So the NeoVIM implementation could look like goto_nextseverities: { ERROR, HINT }}
which would mean first find errors, then hints.

```lua
local M = {}

M.pos_equal = function (p1, p2)
  local r1, c1 = unpack(p1)
  local r2, c2 = unpack(p2)
  return r1 == r2 and c1 == c2
end

M.goto_error_then_hint = function ()
  local pos = vim.api.nvim_win_get_cursor(0)
  vim.diagnostic.goto_next( {severity=vim.diagnostic.severity.ERROR, wrap = true} )
  local pos2 = vim.api.nvim_win_get_cursor(0)
  if ( M.pos_equal(pos, pos2)) then
    vim.diagnostic.goto_next( {wrap = true} )
  end
end

return M


```

**Reply by contrun** [reply]
This works great! Thanks. The only problem for this function is sometimes I will get two "No more valid diagnostics to move to" in the modeline (when there is no action required at all).
